turtle_tf2_py/turtle_tf2_py/turtle_tf2_listener.py

```diff
@@ -37,61 +37,78 @@ def __init__(self):
         self.target_frame = self.get_parameter(
             'target_frame').get_parameter_value().string_value
 
-        self._tf_buffer = Buffer()
-        self._tf_listener = TransformListener(self._tf_buffer, self)
+        self.tf_buffer = Buffer()
+        self.tf_listener = TransformListener(self.tf_buffer, self)
 
         # Create a client to spawn a turtle
-        self.client = self.create_client(Spawn, 'spawn')
-
-        # Check if the service is available
-        while not self.client.wait_for_service(timeout_sec=5.0):
-            self.get_logger().info('service not available, waiting again...')
-
-        # Initialize request with turtle name and coordinates
-        # Note that x, y and theta are defined as floats in turtlesim/srv/Spawn
-        request = Spawn.Request()
-        request.name = 'turtle2'
-        request.x = float(4)
-        request.y = float(2)
-        request.theta = float(0)
-        # Call request
-        self.client.call_async(request)
+        self.spawner = self.create_client(Spawn, 'spawn')
+        # Boolean values to store the information
+        # if the service for spawning turtle is available
+        self.service_called = False
+        # if the turtle was successfully spawned
+        self.turtle_spawned = False
 
         # Create turtle2 velocity publisher
-        self.turtle_vel_ = self.create_publisher(Twist, 'turtle2/cmd_vel', 1)
+        self.publisher = self.create_publisher(Twist, 'turtle2/cmd_vel', 1)
 
         # Call on_timer function every second
-        self._output_timer = self.create_timer(1.0, self.on_timer)
+        self.timer = self.create_timer(1.0, self.on_timer)
 
     def on_timer(self):
         # Store frame names in variables that will be used to
         # compute transformations
         from_frame_rel = self.target_frame
         to_frame_rel = 'turtle2'
 
+        if not self.service_called:
+            if not self.spawner.service_is_ready():
+                # Check if the service is ready
+                self.get_logger().info('Service is not ready')
+                return
+
+            # Initialize request with turtle name and coordinates
+            # Note that x, y and theta are defined as floats in turtlesim/srv/Spawn
+            request = Spawn.Request()
+            request.name = 'turtle2'
+            request.x = float(4)
+            request.y = float(2)
+            request.theta = float(0)
+            # Call request
+            self.result = self.spawner.call_async(request)
+            self.service_called = True
+            return
+        elif self.service_called and not self.turtle_spawned:
+            self.get_logger().info(
+                f'Successfully spawned {self.result.result().name}')
```

Comment:
you are assuming here that the service is already done.

```python
elif self.service_called and not self.turtle_spawned and self.result.done():
```

check this [example](https://github.com/ros2/demos/blob/master/demo_nodes_py/demo_nodes_py/services/add_two_ints_client_async.py)

Reply:
https://github.com/ros/geometry_tutorials/pull/58/commits/75a616fe19021b367156b673aa14b26a3212c844